### DISCUSSION TITLE
Can we have different stateQualifiers for different windows?

### DISCUSSION THREAD

**Discussion by BurningTreeC** [TiddlyWiki/TiddlyWiki5#5787] [Ideas]
Currently the `getStateQualifier` method produces the same state qualifiers for the `root window` and a window where I opened the `$:/core/ui/PageTemplate` tiddler

I'm thinking that it might be possible to add a suffix to the result of `getStateQualfier` based on in which window the concerned widget is

**Comment by Jermolene** [comment]
Hi @BurningTreeC alternately, could we just arrange an extra transclusion in external windows so that we would automatically generate different state qualifiers?

**Reply by BurningTreeC** [reply]
Hi @Jermolene, yes that would be another idea. But wouldn't then be the state qualifiers the same for two external windows?

**Reply by Jermolene** [reply]
The idea would be to get a value into the transclusion chain that represents the unique identity of the window, to ensure there would be no clashes. For example, we might use a numeric index to identify each window, and wrap the window with a `<$tiddler>` widget that sets that value.

**Reply by BurningTreeC** [reply]
@Jermolene, that would be great. I'm not able to dig deeper into that because I don't know how to wrap the window with a `<$tiddler>` widget and I don't know how to set that numeric index for each window

**Reply by Jermolene** [reply]
> @Jermolene, that would be great. I'm not able to dig deeper into that because I don't know how to wrap the window with a `<$tiddler>` widget 

When we open a new window we construct a widget tree representing the content of the window. The parse tree nodes representing the content is obtained from the `$tw.wiki.parseTiddler(template)` call, and then passed to `$tw.wiki.makeWidget()`. You should be able to instead use `$tw.wiki.makeTranscludeWidget()` with the `recursionMarker` option to add the unique index.

> and I don't know how to set that numeric index for each window

In `$:/core/modules/startup/windows.js` we'd maintain a global counter that we increment and capture for each new window.

**Reply by saqimtiaz** [reply]
We have had requests in the past to be able to close a window (via wikitext) that was previously opened with `tm-open-window`. This would also require saving a reference to each new window that was opened. I mention this as something to keep in mind if refactoring code to do with opening new browser windows.

**Reply by Jermolene** [reply]
> We have had requests in the past to be able to close a window (via wikitext) that was previously opened with `tm-open-window`. This would also require saving a reference to each new window that was opened. I mention this as something to keep in mind if refactoring code to do with opening new browser windows.

Great idea.

**Reply by BurningTreeC** [reply]
@Jermolene @saqimtiaz I still don't know how to do this

**Reply by saqimtiaz** [reply]
@BurningTreeC some very quick thoughts: 


it looks like we already save references to each window object, though they aren't unique as the key is just the tiddler title (the same tiddler could be opened multiple times, especially via different templates). The combination of tiddler + template could perhaps be considered unique.

https://github.com/Jermolene/TiddlyWiki5/blob/4f33d2f35cd2c34ca5edfcf062f23bd1ca7ef177/core/modules/startup/windows.js#L41-L47

If we use a unique identifier for each window (perhaps optionally provided by the user) we can use that identifier or a hash of it for the transclusion chain. We could then potentially allow a wikitext way to close a window where the user specifies the identifier and it used to identify the saved reference to the window object in JavaScript.

**Reply by BurningTreeC** [reply]
I cannot do anything with these informations. I think I'll give up on this

**Reply by saqimtiaz** [reply]
@BurningTreeC first off, no pressure here. If you do want to take a crack at this and learn a bit more about the TW internals, I would be happy to try to guide you.

For example, consider the following code from window.js:
https://github.com/Jermolene/TiddlyWiki5/blob/4f33d2f35cd2c34ca5edfcf062f23bd1ca7ef177/core/modules/startup/windows.js#L69-L74

To wrap this inside a tiddler widget, the following is one very straight forward approach. We create a tiddler widget and then use that as the parent for the content widget that we create. 

```javascript
srcDocument.head.insertBefore(styleElement,srcDocument.head.firstChild); 
// Create a tiddler widget
var tiddlerParseTree = {
	"type": "tiddler",
	"attributes": {
		"tiddler": {
			"name": "tiddler",
			"type": "string",
			"value": title+template
		}
	},
	"tag": "$tiddler",
	children: []
};
var tiddlerWidgetNode = new tiddlerWidget.tiddler(tiddlerParseTree, {
	wiki: $tw.wiki,
	document: srcDocument,
	parentWidget: $tw.rootWidget
});
tiddlerWidgetNode.render(srcDocument.body,srcDocument.body.firstChild);
// Render the text of the tiddler
var parser = $tw.wiki.parseTiddler(template),
	widgetNode = $tw.wiki.makeWidget(parser,{document: srcDocument, parentWidget: tiddlerWidgetNode, variables: variables});
widgetNode.render(srcDocument.body,srcDocument.body.firstChild);
 // Function to handle refreshes
```

Note how I have used the title+template as a unique identifier. This pre-supposes that there will never be more than one window open for the same title and template. This could be made even more unique by appending with a counter for the open windows, or  by using a user supplied identifier.

(A user supplied identifier for windows would also allow closing windows more easily in response to a widget message.)

**Reply by saqimtiaz** [reply]
PS: this pre-supposes something like this at the top of the file to import the tiddler widget:

`var tiddlerWidget = require("$:/core/modules/widgets/tiddler.js");`

**Comment by BurningTreeC** [comment]
@Jermolene @saqimtiaz , I'm not sure anymore if this is needed. If I put a `<<qualify>>` macro on top of `$:/core/ui/PageTemplate` I already get different qualifiers for the PageTemplate rendered in different windows

**Reply by AnthonyMuscio** [reply]
I would like to use this a window identification to be able to close such windows, but in a similar vein I wish we had a session ID, and if it were possible a browser ID that we could interrogate as variables. The reasons are many and has the same reasoning as to why sessions are needed on most sites.

**Reply by AnthonyMuscio** [reply]
I can use the browser sniff plugin to get a browser type Firefox/chrome, But I still can't get a unique, session ID, perhaps a random number that I save in local storage? lost on tab close?

If we had all these window, browser, session IDs we could enforce and manage single file wikis so they will not save over each other or windows we open in a wiki, we can also close from the wiki, very useful for Dashboards, bookmark managers and other tiddlywiki applications that need an "out of bound windows".

"state qualifiers" for all imaginable needs.

Just building blocks.

**Comment by AnthonyMuscio** [comment]
I just stumbled upon this which opens a window and seems to close it "remotely"

[Window close() Method](https://www.w3schools.com/jsref/met_win_close.asp)

```
function openWin() {
  myWindow = window.open("", "myWindow", "width=200, height=100");   // Opens a new window
}

function closeWin() {
  myWindow.close();   // Closes the new window
}
```

But perhaps this is already well known.
